sw/device/lib/drivers/gpio/gpio.c

```diff
@@ -1,213 +1,525 @@
-                                bool *state) {
-  if (gpio == NULL || state == NULL) {
-    return kGpioBadArg;
-  }
+gpio_result_t gpio_dis_input_sampling (gpio_pin_number_t pin)
+{
+    if (pin > (MAX_PIN-1) || pin < 0)
+        return GpioPinNotAcceptable;
+    gpio_peri->GPIO_EN0 = bitfield_write(gpio_peri->GPIO_EN0,
+                                         MASK_1b, pin, GPIO_EN__DISABLED);
+    return GpioOk;
+}
 
-  *state = mmio_region_get_bit32(gpio->params.base_addr,
-                                 GPIO_GPIO_IN_REG_OFFSET, pin);
+gpio_result_t gpio_reset (gpio_pin_number_t pin)
+{
+    if (pin > (MAX_PIN-1) || pin < 0)
+        return GpioPinNotAcceptable;
+
+    gpio_intr_set_mode (0);
+    gpio_set_mode (pin, GpioModeIn);
+    gpio_dis_input_sampling (pin);
+    gpio_peri->GPIO_CLEAR0 = bitfield_write(gpio_peri->GPIO_CLEAR0,
+                                         MASK_1b, pin, GPIO_CLEAR__REMOVE_MASK);
+    gpio_peri->GPIO_SET0 = bitfield_write(gpio_peri->GPIO_SET0,
+                                         MASK_1b, pin, GPIO_SET__REMOVE_MASK);
+    gpio_intr_dis_all(pin);
+    gpio_intr_clear_stat(pin);
+    return GpioOk;
+}
 
-  return kGpioOk;
+gpio_result_t gpio_reset_all (void)
+{
+    gpio_peri->GPIO_MODE0 = 0;
+    gpio_peri->GPIO_MODE1 = 0;
+    gpio_peri->GPIO_EN0 = 0;
+    gpio_peri->GPIO_CLEAR0 = 0;
+    gpio_peri->GPIO_SET0 = 0;
+    gpio_peri->GPIO_TOGGLE0 = 0;
+    gpio_peri->INTRPT_RISE_EN0 = 0;
+    gpio_peri->INTRPT_FALL_EN0 = 0;
+    gpio_peri->INTRPT_LVL_HIGH_EN0 = 0;
+    gpio_peri->INTRPT_LVL_LOW_EN0 = 0;
+    gpio_peri->INTRPT_STATUS0 = 0xFFFFFFFF;
 }
 
-gpio_result_t gpio_write(const gpio_t *gpio, gpio_pin_t pin,
-                                 bool state) {
+gpio_result_t gpio_read (gpio_pin_number_t pin, bool *val)
+{
+    if (pin > (MAX_PIN-1) || pin < 0)
+        return GpioPinNotAcceptable;
+    if (bitfield_read(gpio_peri->GPIO_IN0, MASK_1b, pin) == GPIO_IS_SET)
+        *val = true;
+    else
+        *val = false;
+    return GpioOk;
+}
 
-  if (gpio == NULL) {
-    return kGpioBadArg;
-  }
+//todo: check it
+gpio_result_t gpio_toggle (gpio_pin_number_t pin)
+{
+    if (pin > (MAX_PIN-1) || pin < 0)
+        return GpioPinNotAcceptable;
+    if (bitfield_read(gpio_peri->GPIO_OUT0, MASK_1b, pin) == GPIO_IS_SET)
+    {
+        gpio_peri->GPIO_OUT0 = bitfield_write(gpio_peri->GPIO_OUT0, 
+            MASK_1b, pin, GPIO_IS_RESET);
+    }
+    else
+    {
+        gpio_peri->GPIO_OUT0 = bitfield_write(gpio_peri->GPIO_OUT0, 
+            MASK_1b, pin, GPIO_IS_SET);
+    }
+    return GpioOk;
+}
 
-  uint32_t reg_value = mmio_region_read32(gpio->params.base_addr, GPIO_GPIO_OUT_REG_OFFSET);
-  reg_value = bitfield_bit32_write(reg_value, pin, state);
-  mmio_region_write32(gpio->params.base_addr, GPIO_GPIO_OUT_REG_OFFSET, reg_value);
+gpio_result_t gpio_write (gpio_pin_number_t pin, bool val)
+{
+    if (pin > (MAX_PIN-1) || pin < 0)
+        return GpioPinNotAcceptable;
+    if (val == GPIO_IS_SET)
```

Comment:
you can avoid the if else, use the val argument, a bool is expanded to integer (1 for true, 0 for false)